ARCHITECTURE.md

```diff
@@ -0,0 +1,71 @@
+# Architecture
+
+Rojo is a rather large project with a bunch of moving parts. While it's not too complicated in practice, it tends to be overwhelming because it's a fair bit of Rust and not very clear where to begin.
+
+This document is a "what the heck is going on" level view of Rojo and the codebase that's written to make it more reasonable to jump into something. It won't go too into depth on *how* something is done, but it will go into depth on *what* is being done.
+
+## Overarching
+
+Rojo is divided into two main pieces: the server and the plugin. The server is what's ran on your computer (whether it be via the terminal or the visual studio code extension), with the plugin serving as its main client.
+
+When serving a project, the server gathers data on all of the files in that project, puts it into a nice format, and then sends it to the plugin. Then, when something changes on the file system, it does the same thing for only the changed files and sends them to the plugin.
+
+When it receives a patch (whether it be the initial patch or any subsequent ones), the plugin reads through it and attempts to to apply it. Any sugar (the patch visualizer, as an example) happens on top of the patches received from the server.
```

Comment:
```suggestion
When the plugin receives a patch from the server (whether it be the initial patch or any subsequent ones), the plugin reads through the patch and attempts to to apply the changes described by it. Any sugar (the patch visualizer, as an example) happens on top of the patches received from the server.
```
Too many ambiguous uses of "it".

Reply:
Also, this section is mildly misleading, as the initial patch is actually computed by the plugin and not received from the server. That's a bit of complexity that is explained in the plugin section, so I think keeping it simple up here is fine. The application process is the same, and that's what the focus is here.

Reply:
> as the initial patch is actually computed by the plugin and not received from the server

I sort of assumed we were sending an initial patch file because we compute one when a `ServeSession` is initialized in the CLI as part of building the `RojoTree`. I imagine this is just a semantics difference though: regardless of what's sent, the initial patch would have to be computed by the plugin to bring it to parity with the server.